### ISSUE TITLE
request: filter autorepeat in "key" objects

### ISSUE THREAD

**Issue by porres** [pure-data/pure-data#945]
As discussed in https://github.com/pure-data/pure-data/issues/729

We can filter autorepeat of keys in Pd only by setting things in system configuration (at least for macOS), even though other software are able to ignore autorepeat in the case system preference enables it. I was hoping we could do the same to Pd and filter autorepeat from system configuration.

This could be done with a second optional argument.

This is really useful for implementing a musical keyboard interface from you computer keys, for instance.

**Comment by megrimm** [comment]
@porres 

Yeah I run into this each summer when working with students building DIY buttons for triggering events especially "dance pad" type buttons. 

When they are standing on the pad, for example, the [key] just endlessly "repeats" so then I have them change System Preferences -> Keyboard -> Key Repeat = OFF .... THEN restart Pd but it would be nice to not have to remember this each year and just have it all work fine in Pd with [keyup] sending key number when key released to reset [once].

**Comment by megrimm** [comment]
@porres yeah the easiest thing i could figure out is just using [timer] to time [keyup] on/off and then adding a [delay 50] or whatever (as long as its great than [timer] output), right after the [keyup].

**Comment by porres** [comment]
purr data deals with it, so I'd start by checking it out https://github.com/agraef/purr-data/search?p=2&q=autorepeat&unscoped_q=autorepeat

**Comment by megrimm** [comment]
@porres add:

`exec defaults write $::pd_guiprefs::domain KeyRepeat 0`

at line 150 in pd_guiprefs.tcl
seems to work for me to just universally turn key repeating off so [keyup] only sends a bang after key is released.

**Comment by megrimm** [comment]
ugh... nevermind. this is just command line. i jumped gun on it working... but i think this might be general idea?

**Comment by danomatika** [comment]
Why not just use `[change]`? I say that after spending *lots* of time just getting key release to work again on macOS, then deciding to throw in the key repeat system preference handling.

**Comment by danomatika** [comment]
Nevermind, I see it won't work as it's successful down/up event pairs. We can just disable the key repeat part in the TK patch or, if there is a way to know which event is a report in Tk, ignore it in the pd gui. OTOH, I still find it useful when editing patches, but maybe its not worth it for a flood of events which have to be handled in different ways depending on the context.

**Comment by danomatika** [comment]
@megrimm Digging into the key-related defaults values...

To read the global values, ie. what the sliders control in System Preferences -> Keyboard:

~~~
# value of each key
defaults read NSGlobalDomain InitialKeyRepeat
defaults read NSGlobalDomain KeyRepeat

# or show all values with "Key" in the name
defaults read NSGlobalDomain | grep Key
~~~

The `KeyRepeat`defaults key is the "Key Repeat" slider value and is *not* a boolean aka not true/false, 1/0. I see values of:

* Slow: 120
* Fast: 2
* Never: the last value is kept & `InitialKeyRepeat` is changed to 300000

The `InitialKeyRepeat` defaults key is the "Delay Until Repeat" slider value. I see values of:

* Long: 120
* Short: 15

When you choose the "Never" setting for the "Key Repeat" slider, it look like it saves the `InitialKeyRepeat` to a `InitialKeyRepeat_Level_Saved` key then sets `InitialKeyRepeat` to 300000. If that value is in ms, then the key probably won't repeat for 5 minutes or never if the system ignores long values.

I've tried a number of combinations of `KeyRepeat` and `InitialKeyRepeat` defaults values on macOS 10.14, but it doesn't seem this key is taken into account on a per-application level. If it did, then setting the following would work, but it unfortunately doesn't for me:

~~~
defaults write org.puredata.pd.pd-gui InitialKeyRepeat -int 300000
~~~

@umlaeute How is key repeat handled on other systems? If it isn't, then we should probably simply not have repeat on macOS builds either.

**Comment by umlaeute** [comment]
on linux (xfce4, Tcl/Tk-8.6.10), i get keydown+keyup events while holding a key:

~~~tcl
bind . <KeyPress> "puts {KeyPress %K %s}"
bind . <KeyRelease> "puts {KeyRelease %K %s}"
~~~

gives me (while keeping <kbd>a</kbd> pressed):
~~~
KeyPress a 0
KeyRelease a 0
KeyPress a 0
KeyRelease a 0
KeyPress a 0
KeyRelease a 0
KeyPress a 0
KeyRelease a 0
~~~
